### PR TITLE
Meson build system

### PR DESCRIPTION
**Pull request by Salamandar**
This merge request provides the port to the Meson build system.

If you don't know Meson : it's a modern build system, thought as a replacement for the old Autotools. 
It's based on Ninja, a Make "replacement" that's way faster.
Its description language is declarative, meaning there's no hidden behaviour / variables as it's usually the case with Autotools. It's also way easier to read and debug than autoconf.

### CODE REVIEW

**Comment by tristanheaven** [comment]
This is harder than you thought it would be? :)

I'm not sure I like Meson but I'll have to merge this anyway when you're done because you've put so much effort in. Thanks!

**Comment by Salamandar** [comment]
@tristanheaven Ahah :D No, all those force pushes only are to fix the CI (windows and travis), there are some small subtilities.

Don't feel forced to merge this ^^ But you should check Meson, it's really awesome to work with !

**Comment by Salamandar** [comment]
Yay ! Now to fix the code coverage nonsense.

**Comment by ItzSwirlz** [comment]
> I'm not sure I like Meson

You can have it as a second option-keep autotools and meson files like MATE does

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+# Todo check which options are used by Meson
+compiler_options = [
+    '-fno-common',
+    '-fvisibility=hidden',
+    '-pedantic',
+    '-Waggregate-return',
+    '-Wbad-function-cast',
+    '-Wcast-align',
+    '-Werror=implicit-function-declaration',
+    '-Werror=unknown-warning-option',
```
Comment (tristanheaven) [inline meson.build:62]:
```suggestion
```
meson seems to do this automatically (https://github.com/mesonbuild/meson/commit/b289547971f3bf11eae6da195ed4ed014aa57fdd)

File: src/nautilus/meson.build [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            libutil,
+        ],
+        dependencies: [
+            extension.get('dep'),
+        ],
+        install: true,
+        install_dir: extension.get('dir'),
+    )
+
+    # -export-symbols-regex ^caja_module_
```
Comment (tristanheaven) [inline src/nautilus/meson.build:66]: `-export-symbols-regex` is for exposing only the necessary interface to the file manager, e.g, caja expects to find these functions:

caja_module_initialize
caja_module_shutdown
caja_module_list_types

File: meson.build (hunk 1 above)
Comment (Salamandar) [reply meson.build:62]: Yes, I was not 100% sore so I let it, but now that you pointed me to the code, I'll remove it :)

File: src/nautilus/meson.build (hunk 2 above)
Comment (Salamandar) [reply src/nautilus/meson.build:66]: OK, I'll work on it.

Comment (Salamandar) [reply src/nautilus/meson.build]: Done ! Meson does not use libtool so we need to use mapfiles. Fortunately this is a really standard thing.

**Comment by tristanheaven** [comment]
Thanks @Salamandar. I think we can merge the rest of this now (if github allows).

**Comment by Salamandar** [comment]
Amazing ! Now you "just" have to wait for maintainers to use Meson :)

**Pull request merged**